content/en/BestPractices/akhan_input_validation.md

```diff
@@ -0,0 +1,155 @@
+  const product = await QuantityModel.findOne({ where: { ProductId: id } })
+  if (!product) {
+    throw new Error('No such product found!')
+  }
+
+  if (!product.limitPerUser || (product.limitPerUser && product.limitPerUser >= quantity) || security.isDeluxe(req)) {
+    if (product.quantity >= quantity) { // enough in stock?
+      next()
+    } else {
+      res.status(400).json({ error: res.__('We are out of stock! Sorry for the inconvenience.') })
+    }
+  } else {
+    res.status(400).json({ error: res.__('You can order only up to {{quantity}} items of this product.', { quantity: product.limitPerUser.toString() }) })
+  }
+}
+```
+
+### Analysis
+
+The OWASP Juice Shop uses an [API
+Post](https://github.com/juice-shop/juice-shop/blob/bd7e8f3c7af1a8f38ba4fa866000136cb472449a/server.ts#L370)
+to call a
+[`quantityCheck()`](https://github.com/juice-shop/juice-shop/blob/bd7e8f3c7af1a8f38ba4fa866000136cb472449a/routes/basketItems.ts#L90)
+function.
+
+It checks the quantity in the request payload with validation logic, using
+`product.limitPerUser`. However, it does not check if the quantity exceeds the:
+
+- Limit per user
+- Limit per order
+
+It even skips the check for deluxe users (`security.isDeluxe(req)`). However, it
+does *not* check if the quantity is positive.
+
+This small mistake is costly. Users can freely input negative values for
+`quantity`. If an attacker adds `-100` to their basket, the Juice Shop adds 100
+times the value of the item to their wallet.
+
+### Impact
+
+Without such checks, this error could lead to substantial business losses.
+
+### Prevention
+
+In the following code sample, we include code that:
+
+- Checks if the quantity is greater than zero (`quantity > 0`)
+- Runs a boolean check for decimal integers (`Number.isInteger(quantity, 10)`)
+- Returns an HTTP 400 error if the noted checks fail
+
+```ts
+async function quantityCheck (req: Request, res: Response, next: NextFunction, id: number, quantity: number) {
+  const product = await QuantityModel.findOne({ where: { ProductId: id } })
+  if (!product) {
+    throw new Error('No such product found!')
+  }
+
+ // Check if the quantity is a positive decimal integer value, if not send a `400` error response
+ if(!Number.isInteger(quantity, 10) || quantity <= 0) {
+      res.status(400).json({ error: res.__('Invalid Quantity of items.') })
+ }
+
+  if (!product.limitPerUser || (product.limitPerUser && product.limitPerUser >= quantity) || security.isDeluxe(req)) {
+    if (product.quantity >= quantity) { // enough in stock?
+      next()
+    } else {
+      res.status(400).json({ error: res.__('We are out of stock! Sorry for the inconvenience.') })
+    }
+  } else {
+    res.status(400).json({ error: res.__('You can order only up to {{quantity}} items of this product.', { quantity: product.limitPerUser.toString() }) })
+  }
+}
+```
+
+## Best Practices
+
```

Comment:
Two more points I think I missed:

- Always check if the user input matches the right format.
- Always ensure that the user input fits the given business context
  - A user should only enter a number in a quantity field, but the user should not be able to enter negative values there.
  - There should be sensible minimum and maximum values for a given quantity.

Reply:
I love examples!

I've rewritten what you've suggested slightly, and added it in https://github.com/cobalthq/cobalt-product-public-docs/pull/154/files#r946719506